### ISSUE TITLE
Make Timestamp delegate eql? to @time.

### ISSUE THREAD

**Issue by guyboertje** [elastic/logstash#4324]
This fails in RSpec.

``` ruby
its_xmas = Time.utc(2015, 12, 25, 0, 0, 0)
expect(LogStash::Timestamp.new(its_xmas)).to eql(LogStash::Timestamp.new(its_xmas))
```

**Comment by untergeek** [comment]
That is an epic face palm there.  What's going on?

**Comment by jsvd** [comment]
to make this more explicit:

```ruby
LogStash::Timestamp.new(its_xmas).eql?(LogStash::Timestamp.new(its_xmas))
=> false
```

**Comment by jsvd** [comment]
though I should note that == works fine:

```ruby
LogStash::Timestamp.new(its_xmas) == LogStash::Timestamp.new(its_xmas)
 => true
```

**Comment by jordansissel** [comment]
+1 on doing this.

**Comment by purbon** [comment]
We should certainly fix this. 

@jsvd I wonder why ```==``` actually works, I understand this checks the "value"?  right? also the operator are not forwarded to time, interesting. Do you know why it works?

**Comment by jsvd** [comment]
This still happens in logstash 5.3.1:

```
1.9 :007 > its_xmas = Time.utc(2015, 12, 25, 0, 0, 0)
 => 2015-12-25 00:00:00 UTC 
1.9 :008 > its_xmas.eql?(its_xmas)
 => true 
1.9 :009 > LogStash::Timestamp.new(its_xmas).eql?(LogStash::Timestamp.new(its_xmas))
 => false 
```

**Comment by jordansissel** [comment]
Looking back at this, I think `==` is the only one that matters.

Pasting from the ruby docs: http://ruby-doc.org/core-2.4.1/Object.html#method-i-eql-3F

> The eql? method returns true if obj and other refer to the same hash key. This is used by Hash to test members for equality.

Sounds like `eql?` is only used for Hash keys, and we are unlikely to be using LogStash::Timestamp objects as keys, right?

I think the problem is the usage of `eql` in rspec (anywhere). We should be using `==`. For example:

```
expect(value).to be == whatever
```

**Comment by original-brownbear** [comment]
@jsvd added this to the list of things for me to look into so I did in #7002 :)
I think `.eql?` and `==` should be overrides of one another. As the above link says:

> The eql? method returns true if obj and other refer to the same hash key. This is used by Hash to test members for equality. For objects of class Object, eql? is synonymous with ==. Subclasses normally continue this tradition by aliasing eql? to their overridden == method, but there are exceptions.

Exceptions only apply to numeric types apparently and in that case only to type conversion behavior => probably a good idea to stick with that and prevent future random problems from the accidental use of `eql?` right ?

**Comment by colinsurprenant** [comment]
+1 on having `==` and `eql?` be the same here. Both `==` and `eql?` value equality checks, while `equal?` is a reference equality check.

**Comment by jsvd** [comment]
> Sounds like eql? is only used for Hash keys, and we are unlikely to be using LogStash::Timestamp objects as keys, right?

I agree this will be uncommon but given the small change necessary I think it's worth having it and avoid a surprise in the future.

**Comment by jsvd** [comment]
closed by #7002
